fault/tester_samples.py

```diff
@@ -1,46 +1,131 @@
 import fault
-import random
 
 
-class SRAMTester(fault.Tester):
-    def __init__(self, circuit, *args, n_trials=100, prob1=0.5,
-                 supply1='vdd', supply0='vss', **kwargs):
+class GenericCellTester(fault.Tester):
+    def __init__(self, circuit, *args, n_trials=100, supply0='vss',
+                 supply1='vdd', **kwargs):
         # call super constructor
         super().__init__(circuit, *args, **kwargs)
 
         # save settings
-        self.n_trials = n_trials
-        self.prob1 = prob1
-        self.supply1 = supply1
         self.supply0 = supply0
+        self.supply1 = supply1
+        self.n_trials = n_trials
 
         # define the test
         self.define_test()
 
-    def define_test(self):
-        # initialize pin values
-        self.poke(self._circuit.wl, 0)
-        self.poke(self._circuit.lbl, 0)
-        self.poke(self._circuit.lblb, 0)
-        if hasattr(self.circuit, self.supply1):
-            self.poke(getattr(self._circuit, self.supply1), 1)
-        if hasattr(self.circuit, self.supply0):
-            self.poke(getattr(self._circuit, self.supply0), 0)
+    def has_pin(self, name):
+        return hasattr(self._circuit, name)
+
+    def poke(self, name, *args, **kwargs):
+        super().poke(getattr(self._circuit, name), *args, **kwargs)
+
+    def poke_optional(self, name, *args, **kwargs):
+        if self.has_pin(name):
+            self.poke(name, *args, **kwargs)
+
+    def expect(self, name, *args, **kwargs):
+        super().expect(getattr(self._circuit, name), *args, **kwargs)
+
+    def define_init(self):
+        self.poke_optional(self.supply1, 1)
+        self.poke_optional(self.supply0, 0)
+
+    def define_trial(self):
+        pass
```

Comment:
You could make this an abstractmethod (need to use ABCMeta too, see https://docs.python.org/3/library/abc.html#abc.abstractmethod) to make sure that derived classes implement it.

Reply:
Thanks, will include this in the next PR